test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java

```diff
@@ -0,0 +1,273 @@
+        // Thread is parked, for example: LockSupport.park, LockSupport.parkUtil and LockSupport.parkNanos.
+        TestStatus status = new TestStatus("JVMTI_THREAD_STATE_PARKED");
+        CountDownLatch ready = new CountDownLatch(1);
+        Thread vthread = createPinnedVThread(() -> {
+            ready.countDown();
+            LockSupport.park(Thread.currentThread());
+        });
+        vthread.start();
+        ready.await();
+        Thread.sleep(500); // wait some time to ensure the thread has reached waiting state
+        // don't test interrupt() - it causes thread state change for parked thread
+        // even if it's suspended
+        testThreadStates(vthread, false,
+                JVMTI_THREAD_STATE_WAITING | JVMTI_THREAD_STATE_WAITING_INDEFINITELY | JVMTI_THREAD_STATE_PARKED);
+        // allow test thread to finish
+        LockSupport.unpark(vthread);
+        status.print();
+    }
+
+    static void inNative() throws Exception {
+        TestStatus status = new TestStatus("JVMTI_THREAD_STATE_IN_NATIVE");
+        Thread vthread = createPinnedVThread(() -> {
+            waitInNative();
+        });
+        vthread.start();
+        while (!waitInNativeReady) {
+            Thread.sleep(50);
+        }
+        testThreadStates(vthread, true,
+                JVMTI_THREAD_STATE_RUNNABLE | JVMTI_THREAD_STATE_IN_NATIVE,
+                0);
+        status.print();
+    }
+
+
+    public static void main(String[] args) throws Exception {
+        runnable();
+        /* "waiting" test cases fail due JDK-8310584
+        blockedOnMonitorEnter();
+        waiting(false);
+        waiting(true);
+        sleeping();
+        parked();
+        */
+        inNative();
+
+        int errCount = getErrorCount();
+        if (errCount > 0) {
+            throw new RuntimeException("Test failed, " + errCount + " errors");
+        }
+    }
+
+    private static Thread createPinnedVThread(Runnable runnable) {
+        final Object syncObj = new Object();
+        return Thread.ofVirtual().unstarted(() -> {
+            synchronized (syncObj) {
+                runnable.run();
+            }
+        });
+    }
+
+    // Tests thread states (vthread and the carrier thread).
+    // expectedStrong specifies value which must be present in vthreat state;
+    // expectedWeak is a combination of bits which may be set in vthreat state
+    // (at least one of the bit must set, but not all).
+    private static native void testThread(Thread vthread,
+                                          boolean testInterrupt,
+                                          int expectedStrong, int expectedWeak);
+    private static native int getErrorCount();
+
+    private static boolean waitInNativeReady = false;
+
+    // Sets waitInNativeReady static field to true
+    // and then waits until endWait() method is called.
+    private static native void waitInNative();
+    // Signals waitInNative() to exit.
+    private static native void endWait();
```

Comment:
Q: Where is this native method invoked?

Reply:
Good catch. Test thread inNative testcase never exits, it's terminated on the test exit.
Fixed.